implemented ISO 8601 date duration parsing

SATVIKsynopsis opened the pull request:
<!--
Thank you for your pull request to ICU4X!

Reminder: try to use [Conventional Comments](https://conventionalcomments.org/) to make comments clearer.

Please see https://github.com/unicode-org/icu4x/blob/main/CONTRIBUTING.md for general
information on contributing to ICU4X.
-->
Fixes #7030

This change adds support for parsing ISO 8601 date only duration strings into `DateDuration`. The implementation validates the overall structure, rejects time components and duplicate units, and correctly handles negative values and numeric overflow.

sffc reviewed (commented):
Hello again! Please add unit tests and docs tests. Cover as many edge cases as you can think of. I can't review this further until it is properly tested.

SATVIKsynopsis commented:
I have added unit tests covering valid iso 8601 date only durations as well as common invalid cases and also added as many edge cases as i can think of.

SATVIKsynopsis commented:
All CI checks pass now.

SATVIKsynopsis commented:
thanks for pointing that out and very happy to answer this.

ixdtf is a general iso 8601 parser and supports a wider set of duration forms which also include time components.

`DateDuration` is intentionally a date only type like Y/M/W/D with calendar specific behavior and error handling. using ixdtf here would still require extra validation and mapping to enforce those constraints.

for that reason, this parser is intentionally made instead of reusing the ixdtf parser.

SATVIKsynopsis commented:
i have resolved the nitpick and removed the unnecessary unstable feature gates from the doctests.

SATVIKsynopsis commented:
Hi. Just checking if anything else is required from my side. Thanks

robertbastian commented on components/calendar/src/duration.rs:
```diff
@@ -166,6 +254,113 @@ impl DateDuration {
+            while ch.is_ascii_digit() {
+                has_digits = true;
+
+                let digit = ch as i64 - '0' as i64;
```
```suggestion
            while let Some(digit) = ch.to_digit(10) {
                has_digits = true;
```

SATVIKsynopsis commented:
I updated `try_from_utf8` to use slice based pattern matching instead of indexing so as to avoid panics and clippy warnings. i also moved the main documentation to `try_from_str` from `try_from_utf8`  as suggested.

sffc reviewed (approved):
Let's get this landed. Thank you!

SATVIKsynopsis commented:
Thank you for the review and guidance. Really appreciate it.
